validate/error.go

```diff
@@ -0,0 +1,69 @@
+package validate
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// ComplianceLevel represents the OCI compliance levels
+type ComplianceLevel int
+
+const (
+	ComplianceOptional ComplianceLevel = iota
+	ComplianceMay
+	ComplianceRecommended
+	ComplianceShould
+	ComplianceShouldNot
+	ComplianceShall
+	ComplianceShallNot
+	ComplianceRequired
+	ComplianceMustNot
+	ComplianceMust
+)
+
+// OCIErrorCode represents the compliance content
+type OCIErrorCode int
+
+const (
+	DefaultFilesystems OCIErrorCode = iota
+)
+
+// OCIError represents an error with compliance level and OCI reference
+type OCIError struct {
+	Level     ComplianceLevel
+	Reference string
+	Err       error
+}
+
+//FIXME: change to tagged spec releases
+const referencePrefix = "https://github.com/opencontainers/runtime-spec/blob/master/"
+
+var ociErrors = map[OCIErrorCode]OCIError{
+	DefaultFilesystems: OCIError{Level: ComplianceShould, Reference: "config-linux.md#default-filesystems"},
+}
+
+// ParseLevel takes a string level and returns the OCI compliance level constant
+func ParseLevel(level string) ComplianceLevel {
+	switch strings.ToUpper(level) {
+	case "SHOULD":
+		return ComplianceShould
+	case "MUST":
+		return ComplianceMust
```

Comment:
This doesn't seem quite right.  [RFC 2119][1] has the following levels:

* MUST-level (MUST, MUST NOT, REQUIRED, SHALL, SHALL NOT).
* SHOULD-level (SHOULD, SHOULD NOT, RECOMMENDED, NOT RECOMMENDED).
* MAY-level (MAY, OPTIONAL).

You've [listed yours in the inverse order from the RFC][2] (and left off [NOT RECOMMENDED][3]?), and that means that your `ComplianceMustNot` is less than your `ComplianceMust`, even though both should be in the MUST-level.  I recommend either:

a. Reorder your constants to put `ComplianceShould` at the bottom of the SHOULD-level constants and `ComplianceMust` at the bottom of the MUST-level constants.
b. Add a function that maps the constants to their level (e.g. `ComplianceLevel(ComplianceShallNot) → ComplianceMust`) and use that when deciding whether errors are fatal.
c. Adjust this case here to select whatever the lowest constant is for that level (e.g. `"MUST" → ComplianceShall`).


[1]: https://tools.ietf.org/html/rfc2119
[2]: https://github.com/opencontainers/runtime-tools/pull/354/files#diff-02e70c9c804a3b4108260b193fedf2d7R13
[3]: https://tools.ietf.org/html/rfc2119#section-4